frontend/app/src/data/amount_formatter.ts

```diff
@@ -0,0 +1,44 @@
+import { default as BigNumber } from 'bignumber.js';
+
+export class AmountFormatter {
+  private numberPartsRegex = /(%T.*)(%U.*)(%D)/g;
+  private amountRegex = /%T.*%U.*%D/g;
+
+  format(
+    amount: BigNumber,
+    format: string,
+    precision: number,
+    currency?: string
+  ) {
+    /**
+     * Get separators
+     */
+    const numberParts = [...format.matchAll(this.numberPartsRegex)][0];
+    if (!numberParts)
+      throw new Error('Missing format placeholders: %T, %U or %D');
+    const thousandsSeparator = numberParts[1].substring(2);
+    const decimalSeparator = numberParts[2].substring(2);
+
+    /**
+     * Format the amount using BigNumbers
+     */
+    const formattedAmount = amount.toFormat(
+      amount.modulo(1).comparedTo(0) === 0 ? 0 : precision,
+      BigNumber.ROUND_DOWN,
+      {
+        groupSize: 3,
+        groupSeparator: thousandsSeparator,
+        decimalSeparator
+      }
+    );
+    let formattedString = format.replace(this.amountRegex, formattedAmount);
+
+    /**
+     * Then we replace the currency placeholder
+     */
+    formattedString = formattedString.replace(/%C/gm, currency || '');
```

Comment:
nitpick/question: any reason why not doing a return `format.replace(...).replace(...).trim()`